src/auth/auth.service.ts

```diff
@@ -0,0 +1,156 @@
+    return await this.userService.updateUser(user);
+  }
+
+  async signOut(token: string): Promise<User> {
+    const tokenData: any = this.jwtService.decode(token);
+    const {id, email} = tokenData;
+    const user: User = await this.userService.findUser({id, email});
+    user.token = null;
+    return await this.userService.updateUser(user);
+  }
+
+  validateToken(token: string): boolean {
+    try {
+      this.jwtService.verify(token);
+      return true;
+    } catch (err) {
+      throw new HttpException({
+        status: HttpStatus.FORBIDDEN,
+        errors: [err.message],
+      }, HttpStatus.FORBIDDEN);
+    }
+  }
+
+  async validateSignUpFields(userDto: UserDto): Promise<string[] | null> {
+    const errors = [];
+    const {username, email, password} = userDto;
+    const emailError: string = this.checkEmail(email);
+    const passwordError: string = this.checkPassword(password);
+    const usernameError: string = this.checkUsername(username);
+    const userError: string = await this.checkUserExistence(userDto);
+
+    if (emailError) {
+      errors.push(emailError);
+    }
+    if (passwordError) {
+      errors.push(passwordError);
+    }
+    if (usernameError) {
+      errors.push(usernameError);
+    }
+    if (userError) {
+      errors.push(userError);
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async validateSignInFields(userDto: UserDto): Promise<string[] | null> {
+    const errors = [];
+    const {email, password} = userDto;
+    const emailError: string = this.checkEmail(email);
+    const passwordError: string = this.checkPassword(password);
+    const userError: string = await this.checkUser(userDto);
+
+    if (emailError) {
+      errors.push(emailError);
+    }
+    if (passwordError) {
+      errors.push(passwordError);
+    }
+    if (userError && !errors.length) {
+      errors.push(userError);
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async validateSignOutFields(token: string): Promise<string[] | null> {
+    const errors = [];
+    const tokenData: any = this.jwtService.decode(token);
+    const {id, email} = tokenData;
+
+    if (tokenData) {
+      const user: User = await this.userService.findUser({id, email});
+      if (!user) {
+        errors.push('usr does not exist');
+      }
+    }
+
+    return errors.length ? errors: null;
+  }
+
+  async checkUserExistence(userDto: UserDto): Promise<string> {
+    const {email} = userDto;
+    const user: User = await this.userService.findUser({email});
+
+    if (user) {
+      return 'email is registered';
+    }
+  }
+
+  async checkUser(userDto: UserDto): Promise<string> {
+    const {email, password} = userDto;
+    const user: User = await this.userService.findUser({email});
+    const isUserPass: boolean = await bcrypt.compare(password, user.password);
+
+    if (!user || !isUserPass) {
+      return 'invalid email or password';
+    }
+  }
+
+  checkUsername(username: string): string {
```

Comment:
Those independent functions (only logic without dependencies) could extract to separate service